### PR TITLE
feat: support collect denormalization errors

### PR DESCRIPTION
**Pull request by natacha-h**
| Q             | A
| ------------- | ---
| Branch?       | main
| Tickets       | #4793
| License       | MIT
| Doc PR        | [api-platform/docs#1643](https://github.com/api-platform/docs/pull/1643)

Related to https://github.com/symfony/symfony/pull/21239 and https://github.com/symfony/symfony/pull/42502.

Add property `collectDenormalizationErrors` in Operation (set to `false` by default), setting this property to `true`will enable collecting denormalization type errors to return them as constraint violations.

### CODE REVIEW

File: features/main/validation.feature [hunk 1]
```diff
@@ -72,3 +72,38 @@ Feature: Using validations groups
...
+        },
+        {
+          "propertyPath": "bar",
+          "message": "The type of the \"bar\" attribute must be \"int\", \"string\" given.",
+          "code": "0"
+        }
+      ]
+    }
+    """
+    And the header "Content-Type" should be equal to "application/ld+json; charset=utf-8"
```
Comment (alanpoulain) [inline features/main/validation.feature:109]: **nitpick:** Maybe this line should be before.

File: src/Serializer/AbstractItemNormalizer.php [hunk 2]
```diff
@@ -419,7 +419,7 @@ protected function setAttributeValue(object $object, string $attribute, mixed $v
      *
      * @throws InvalidArgumentException
      */
-    protected function validateType(string $attribute, Type $type, mixed $value, string $format = null): void
+    protected function validateType(string $attribute, Type $type, mixed $value, string $format = null, array &$context): void
```
Comment (alanpoulain) [inline src/Serializer/AbstractItemNormalizer.php:422]: **issue:** It's a BC break (add a default value?)

File: src/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php [hunk 3]
```diff
@@ -226,7 +226,6 @@ private function normalizeDefaults(array $defaults): array
         foreach ($rc->getConstructor()->getParameters() as $param) {
             $publicProperties[$param->getName()] = true;
         }
-
```
Comment (alanpoulain) [inline src/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php:229]: **nitpick:** Please revert these changes.

File: tests/Fixtures/TestBundle/Entity/DummyWithCollectDenormalizationErrors.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+namespace ApiPlatform\Tests\Fixtures\TestBundle\Entity;
+
+use ApiPlatform\Metadata\ApiResource;
+use ApiPlatform\Metadata\Post;
+use Doctrine\ORM\Mapping as ORM;
+
+/**
+ * DummyWithCollectDenormalizationErrors.
+ */
```
Comment (alanpoulain) [inline tests/Fixtures/TestBundle/Entity/DummyWithCollectDenormalizationErrors.php:22]: **nitpick:** Can be removed.

File: features/main/validation.feature (hunk 1 above)
Comment (soyuka) [reply features/main/validation.feature:109]: I think we check the response so its ok

File: src/Serializer/AbstractItemNormalizer.php (hunk 2 above)
Comment (soyuka) [reply src/Serializer/AbstractItemNormalizer.php:422]: it doesn't have to be a reference also

File: src/Symfony/EventListener/DeserializeListener.php [hunk 5]
```diff
@@ -72,11 +77,28 @@ public function onKernelRequest(RequestEvent $event): void
...
+                'data',
+                $this->serializer->deserialize($request->getContent(), $context['resource_class'], $format, $context)
+            );
+        } catch (PartialDenormalizationException $e) {
+            $violations = new ConstraintViolationList();
+            foreach ($e->getErrors() as $exception) {
+                if (!$exception instanceof NotNormalizableValueException) {
+                    continue;
+                }
+                $message = sprintf('The type of the "%s" attribute must be "%s", "%s" given.', $exception->getPath(), implode(', ', $exception->getExpectedTypes() ?? []), $exception->getCurrentType());
```
Comment (norkunas) [inline src/Symfony/EventListener/DeserializeListener.php:91]: Maybe it would be better to use by default a message like in Symfony Validator componen: `This value should be of type {{ type }}.` so it would be translated if validator and translation components are dependencies of application?

**Comment by alanpoulain** [comment]
@norkunas and @lyrixx (because you made the related PR in Symfony :wink:), it would be great if you could review this PR!

**Comment by alanpoulain** [comment]
**Little reminder**: once merged, it would be great to modify the recipe to enable it by default for all resources.

**Review by soyuka — APPROVED** [review]
nice job!

**Review by norkunas — APPROVED** [review]
Nice :slightly_smiling_face:
